src/shapes/group.class.js

```diff
@@ -226,18 +226,12 @@
      * @param {fabric.Object} object
      * @returns
      */
-    canEnter: function (object) {
+    canEnterGroup: function (object) {
       if (object === this || this.isDescendantOf(object)) {
-        /* _DEV_MODE_START_ */
-        console.warn('fabric.Group: trying to add group to itself, this call has no effect');
-        /* _DEV_MODE_END_ */
-        return false;
+        throw new Error('fabric.Group: trying to add group to itself');
```

Comment:
we have to stop going back and fort on this.
Let's not throw errors for now. swapping all the console.error with throw.error is a search/replace away.

Reply:
Did you read this? https://github.com/fabricjs/fabric.js/pull/7859#discussion_r846754553
If these conditions are met fabric will hit an exception anyways but it won't be clear as to why it happened. That's why I think throwing is a MUST.

Reply:
we can use a console.error, and the exception will throw for us?
So what returning false does if not avoiding this object to enter the group?

Reply:
ok so canEnterGroup is, couldHaveEnteredGroup.
Because is called in the callback via enterGroup.
The solution here is to move canEnterGroup out of the callback, in the add function of the group to gatekeep from entering.

Reply:
opening a PR on this branch

Reply:
> ok so canEnterGroup is, couldHaveEnteredGroup. Because is called in the callback via enterGroup. The solution here is to move canEnterGroup out of the callback, in the add function of the group to gatekeep from entering.

Exactly

Reply:
The advantage of moving it to add/insertAt is clear. That's how it should be. The disadvantage is a drop in perf for large groups because we add an iteration over objects. Unless we manage to change collection logic to run all logic on object in one single iteration